chain/types/tipset_selector.go

```diff
@@ -0,0 +1,162 @@
+		Latest:    TipSetSelector{Tag: &TipSetTags.Latest},
+		Finalized: TipSetSelector{Tag: &TipSetTags.Finalized},
+		Height: func(height abi.ChainEpoch, previous bool, anchor *TipSetAnchor) TipSetSelector {
+			return TipSetSelector{Height: &TipSetHeight{At: height, Previous: previous, Anchor: anchor}}
+		},
+		Key: func(key TipSetKey) TipSetSelector { return TipSetSelector{Key: &key} },
+	}
+
+	// TipSetAnchors represents the predefined set of anchors for tipsets.
+	//
+	// See TipSetAnchor.
+	TipSetAnchors = struct {
+		Latest    *TipSetAnchor
+		Finalized *TipSetAnchor
+		Key       func(TipSetKey) *TipSetAnchor
+	}{
+		Latest:    &TipSetAnchor{Tag: &TipSetTags.Latest},
+		Finalized: &TipSetAnchor{Tag: &TipSetTags.Finalized},
+		Key:       func(key TipSetKey) *TipSetAnchor { return &TipSetAnchor{Key: &key} },
+	}
+)
+
+// TipSetTag is a string that represents a pointer to a tipset.
+// See TipSetSelector.
+type TipSetTag string
+
+// TipSetSelector captures the selection criteria for a tipset.
+//
+// The supported criterion for selection is one of the following:
+//   - Key: the tipset key, see TipSetKey.
+//   - Height: the tipset height with an optional fallback to non-null parent, see TipSetHeight.
+//   - Tag: the tipset tag, either "latest" or "finalized", see TipSetTags.
+//
+// At most, one such criterion can be specified at a time. Otherwise, the
+// criterion is considered to be invalid. See Validate.
+type TipSetSelector struct {
+	Key    *TipSetKey    `json:"key,omitempty"`
+	Height *TipSetHeight `json:"height,omitempty"`
+	Tag    *TipSetTag    `json:"tag,omitempty"`
+}
+
+// Validate ensures that the TipSetSelector is valid. It checks that only one of
+// the selection criteria is specified. If no criteria are specified, it returns
+// nil, indicating that the default selection criteria should be used as defined
+// by the Lotus API Specification.
+func (tss TipSetSelector) Validate() error {
+	var criteria int
+	if tss.Key != nil {
+		criteria++
+	}
+	if tss.Tag != nil {
+		criteria++
+	}
+	if tss.Height != nil {
+		criteria++
+		if err := tss.Height.Validate(); err != nil {
+			return xerrors.Errorf("validating tipset height: %w", err)
+		}
+	}
+	if criteria != 1 {
+		return xerrors.Errorf("exactly one tipset selection criteria must be specified, found: %v", criteria)
+	}
+	return nil
+}
+
+// TipSetHeight is a criterion that selects a tipset At given height anchored to
+// a given parent tipset.
+//
+// In a case where the tipset at given height is null, and Previous is true,
+// it'll select the previous non-null tipset instead. Otherwise, it returns the
+// null tipset at the given height.
+//
+// The Anchor may optionally be specified as TipSetTag, or TipSetKey. If
+// specified, the selected tipset is guaranteed to be a child of the tipset
+// specified by the anchor at the given height. Otherwise, the "finalized" TipSetTag
+// is used as the Anchor.
+type TipSetHeight struct {
+	At       abi.ChainEpoch `json:"at,omitempty"`
```

Comment:
So if you omit `"at"` then you'll default to `0` and `Validate()` is going to be fine with that. This is not terrible, but I wonder whether we should be distinguishing an omitted value from `0` here because `0` is almost certainly not going to be what you want out of the API.